components/glhelper.py

```diff
@@ -0,0 +1,244 @@
+    glEnableClientState(GL_VERTEX_ARRAY)
+    glVertexPointer(3, GL_FLOAT, 0, vertices)
+    glDrawArrays(GL_LINE_STRIP, 0, count)
+    glDisableClientState(GL_VERTEX_ARRAY)
+
+
+def draw_helix_memoized(p, n, r, pitch=1, turns=1.0, sides=64):
+    """Draw helix given point, normal vector, radius, pitch, # turns, and # sides.
+    Uses the memoized function compute_helix to lookup points.
+    """
+    vertices, count = compute_helix(*p, *n, r, pitch, turns, sides)
+
+    # draw vertices
+    glEnableClientState(GL_VERTEX_ARRAY)
+    glVertexPointer(3, GL_FLOAT, 0, vertices)
+    glDrawArrays(GL_LINE_STRIP, 0, count)
+    glDisableClientState(GL_VERTEX_ARRAY)
+
+
+@functools.lru_cache(maxsize=None)
+def compute_circle(p0, p1, p2, n0, n1, n2, r, sides):
+    """Compute vertices of circle given point, normal vector, radius, and # sides.
+    Uses an approximation method to compute vertices versus many trig calls.
+    Memoized for quick lookup.
+    """
+    a, _, n = rotate_basis(n0, n1, n2)
+    theta = 6.28318530717958647692 / sides
+    tangential_factor = math.tan(theta)
+    radial_factor = math.cos(theta)
+
+    x, y, z = r * a[0], r * a[1], r * a[2]
+    count = sides + 1
+    vertices = np.empty(count * 3)
+    for i in range(count):
+        vertices[i*3] = x + p0
+        vertices[i*3 + 1] = y + p1
+        vertices[i*3 + 2] = z + p2
+        tx = (y*n[2] - z*n[1]) * tangential_factor
+        ty = (z*n[0] - x*n[2]) * tangential_factor
+        tz = (x*n[1] - y*n[0]) * tangential_factor
+        x = (x + tx) * radial_factor
+        y = (y + ty) * radial_factor
+        z = (z + tz) * radial_factor
+    return vertices, count
+
+
+@functools.lru_cache(maxsize=None)
+def compute_helix(p0, p1, p2, n0, n1, n2, r, pitch, turns, sides):
+    """Compute vertices of helix given point, normal vector, radius, pitch, # turns, and # sides.
+    Uses an approximation method to compute vertices versus many trig calls.
+    Memoized for quick lookup.
+    """
+    a, _, n = rotate_basis(n0, n1, n2)
+    theta = 6.28318530717958647692 / sides
+    tangential_factor = math.tan(theta)
+    radial_factor = math.cos(theta)
+
+    x, y, z = r * a[0], r * a[1], r * a[2]
+    d0, d1, d2 = n[0]*pitch / sides, n[1]*pitch / sides, n[2]*pitch / sides
+    count = int(sides * turns) + 1
+    vertices = np.empty(count * 3)
+    for i in range(count):
+        vertices[i*3] = x + p0 + d0*i
+        vertices[i*3 + 1] = y + p1 + d1*i
+        vertices[i*3 + 2] = z + p2 + d2*i
+        tx = (y*n[2] - z*n[1]) * tangential_factor
+        ty = (z*n[0] - x*n[2]) * tangential_factor
+        tz = (x*n[1] - y*n[0]) * tangential_factor
+        x = (x + tx) * radial_factor
+        y = (y + ty) * radial_factor
+        z = (z + tz) * radial_factor
+    return vertices, count
+
+
+def draw_circle_trig(p, n, r, sides=36):
+    """Draw circle given point, normal vector, radius, and # sides."""
+    a, b, n = rotate_basis(*n)
+    tau = 6.28318530717958647692

```

Comment:
I see this in multiple functions. would it be better to make it a global variable?

Reply:
I'd rather not deal with global variables, so I think it's fine this way.